src/core/cgroup.c

```diff
@@ -740,19 +746,27 @@ static void cgroup_context_apply(Unit *u, CGroupMask mask, ManagerState state) {
         if ((mask & CGROUP_MASK_MEMORY) && !is_root) {
                 if (cg_unified() > 0) {
                         uint64_t max = c->memory_max;
+                        uint64_t swap_max = c->memory_swap_max;
 
-                        if (cgroup_context_has_unified_memory_config(c))
+                        if (cgroup_context_has_unified_memory_config(c)) {
                                 max = c->memory_max;
-                        else {
+                                swap_max = c->memory_swap_max;
+                        } else {
                                 max = c->memory_limit;
 
                                 if (max != CGROUP_LIMIT_MAX)
                                         log_cgroup_compat(u, "Applying MemoryLimit %" PRIu64 " as MemoryMax", max);
+
+                                swap_max = c->memory_swap_limit;
+
```

Comment:
This translation doesn't make sense to me. memsw of cgroup v1 and swap of cgroup v2 have very different semantics and I don't think it'd be a good idea to translate the two back and forth this way. Given that systemd omits other aspects of memcg on cgroup v1 and that memsw is difficult to wrap one's head around, I'd recommend just implementing swap max for cgroup v2.